scripts/check-links.js

```diff
@@ -0,0 +1,309 @@
+            try {
+                onPage(error, pageURL, brokenLinks);
+            }
+            catch(error) {
+                fail(error);
+            }
+        },
+        end: async () => {
+            try {
+                await onComplete(brokenLinks);
+            }
+            catch (error) {
+                fail(error);
+            }
+        },
+    };
+}
+
+// Handles BLC 'link' events, adding broken links to the running list.
+function onLink(result, brokenLinks) {
+    const source = result.base.resolved;
+    const destination = result.url.resolved;
+
+    if (result.broken) {
+        const reason = result.brokenReason;
+
+        addLink(source, destination, reason, brokenLinks);
+
+        // Always log broken links to the console.
+        logLink(source, destination, reason);
+
+    } else if (process.env.DEBUG) {
+
+        // Log successes when DEBUG is truthy.
+        logLink(source, destination, result.http.response.statusCode);
+    }
+}
+
+// Handles BLC 'page' events. For page scope, we fail immediately, since that means the
+// requested page was not found. Otherwise, we just log the failed URL to the running list
+// of broken links.
+function onPage(error, pageURL, brokenLinks) {
+    if (error) {
+        if (checkScope === "page") {
+            throw new Error(`Failed to retrieve ${pageURL}: ${error.message}.`);
+        }
+        else {
+            addLink(pageURL, pageURL, error.message, brokenLinks);
+        }
+    }
+}
+
+// Handles the BLC 'complete' event, which is raised at the end of a run.
+async function onComplete(brokenLinks) {
+    const filtered = excludeAcceptable(brokenLinks);
+
+    if (filtered.length > 0) {
+
+        // If we failed and a retry count was provided, retry. Note that retry count !==
+        // run count, so a retry count of 1 means run once, then retry once, which means a
+        // total run count of two.
+        if (maxRetries > 0 && retryCount < maxRetries) {
+            retryCount += 1;
+            console.log(`Retrying (${retryCount} of ${maxRetries})...`);
+            checkLinks();
+            return;
+        }
+
+        const list = filtered
+            .map(link => `• <${link.source}|${new URL(link.source).pathname}> → ${link.destination} (${link.reason})`)
+            .join("\n");
+
+        // Post the results to Slack.
+        await postToSlack(
+            "ops-notifications",
+            `Eek! :scream_cat: There are broken links on ${new URL(baseURL).hostname}: \n\n ${list}`
+        );
+
+        throw new Error("Finished, but failed with errors. See the log for details.");
+    }
+}
+
+/**
+    We exclude some links:
+    - Our generated API docs have lots of broken links.
+    - Our available versions page includes links to private repos.
+    - GitHub Edit Links may be broken, because the page might not yet exist!
+    - Our LinkedIn page, for some reason, returns an HTTP error (despite being valid).
```

Comment:
LinkedIn probably does this to prevent scraping, or at least that would be my guess as to why it fails.

Reply:
Yeah, that'd make sense. This rule goes back a ways, but if you [Google around for HTTP 999](https://www.google.com/search?q="HTTP"+"999"+"linkedin") (the status code returned for every request we make to a linkedin.com URL), it's clear LinkedIn would prefer not to be crawled. 🤷